sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+      columnarBatch.close();
+      columnarBatch = null;
+    }
+    if (recordReader != null) {
+      recordReader.close();
+      recordReader = null;
+    }
+  }
+
+  /**
+   * Initialize ORC file reader and batch record reader.
+   * Please note that `setRequiredSchema` is needed to be called after this.
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+      throws IOException, InterruptedException {
+    FileSplit fileSplit = (FileSplit)inputSplit;
+    Configuration conf = taskAttemptContext.getConfiguration();
+    reader = OrcFile.createReader(
+      fileSplit.getPath(),
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath().getFileSystem(conf)));
+
+    Reader.Options options =
+            OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart(), fileSplit.getLength());
+    recordReader = reader.rows(options);
+    totalRowCount = reader.getNumberOfRows();
+  }
+
+  /**
+   * Set required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  public void setRequiredSchema(
+    TypeDescription orcSchema,
+    int[] requestedColIds,
+    StructType requiredSchema,
+    StructType partitionSchema,
+    InternalRow partitionValues) {
+    batch = orcSchema.createRowBatch(DEFAULT_SIZE);
+    assert(!batch.selectedInUse); // `selectedInUse` should be initialized with `false`.
+
+    StructType resultSchema = new StructType(requiredSchema.fields());
+    for (StructField f : partitionSchema.fields())
+      resultSchema = resultSchema.add(f);
+    this.requiredSchema = requiredSchema;
+    this.requestedColIds = requestedColIds;
+
+    int capacity = DEFAULT_SIZE;
+    if (DEFAULT_MEMORY_MODE == MemoryMode.OFF_HEAP) {
+      columnVectors = OffHeapColumnVector.allocateColumns(capacity, resultSchema);
+    } else {
+      columnVectors = OnHeapColumnVector.allocateColumns(capacity, resultSchema);
+    }
+    columnarBatch = new ColumnarBatch(resultSchema, columnVectors, capacity);
+
+    if (partitionValues.numFields() > 0) {
+      int partitionIdx = requiredSchema.fields().length;
+      for (int i = 0; i < partitionValues.numFields(); i++) {
+        ColumnVectorUtils.populate(columnVectors[i + partitionIdx], partitionValues, i);
+        columnVectors[i + partitionIdx].setIsConstant();
+      }
+    }
+
+    // Initialize the missing columns once.
+    for (int i = 0; i < requiredSchema.length(); i++) {
+      if (requestedColIds[i] < 0) {
+        columnVectors[i].putNulls(0, columnarBatch.capacity());
+        columnVectors[i].setIsConstant();
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private boolean nextBatch() throws IOException {
```

Comment:
The nextBatch() method is now so long/complex that JVM may have difficulties optimizing it fully. I'm not sure if this will have real-world performance impact, but splitting it to per-type methods would not hurt readability either.

Reply:
yea, at least we should have 3 methods for repeated data, non-nullable data and nullable data.

Reply:
It's updated like the following now.
```scala
if (fromColumn.isRepeating) {
  putRepeatingValues(batchSize, field, fromColumn, toColumn);
} else if (fromColumn.noNulls) {
  putNonNullValues(batchSize, field, fromColumn, toColumn);
} else {
  putValues(batchSize, field, fromColumn, toColumn);
}
```